Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm

```diff
@@ -0,0 +1,255 @@
+    if ((script[filesKey] && functionWasPassed)
+        || (!functionWasPassed && !script[filesKey])) {
+        *outExceptionString = toErrorString(nil, nil, @"Either 'func' or 'files' must be passed, but not both at the same time.");
+        return false;
+    }
+
+    BOOL scriptContainsArguments = !!(script[argsKey] || script[argumentsKey]);
+    if (scriptContainsArguments && !functionWasPassed) {
+        *outExceptionString = toErrorString(nil, nil, @"The 'args' property must be passed along with the 'func' property.");
+        return false;
+    }
+
+    if (script[filesKey]) {
+        NSArray *files = script[filesKey];
+        if (!files.count) {
+            *outExceptionString = toErrorString(nil, filesKey, @"at least one file must be passed.");
+            return false;
+        }
+    }
+
+    if (NSString *world = script[worldKey]) {
+        if (![world isEqualToString:isolatedWorld] && ![world isEqualToString:mainWorld]) {
+            *outExceptionString = toErrorString(nil, worldKey, @"it must either have the value of 'ISOLATED' or 'MAIN'.");
+            return false;
+        }
+    }
+
+    return true;
+}
+
+bool WebExtensionAPIScripting::validateTarget(NSDictionary *targetInfo, NSString **outExceptionString)
+{
+    static NSArray<NSString *> *requiredKeys = @[
+        tabIDKey,
+    ];
+    static NSDictionary<NSString *, id> *keyTypes = @{
+        frameIDsKey: @[ NSNumber.class ],
+        tabIDKey: NSNumber.class,
+        allFramesKey: @YES.class,
+    };
+
+    if (!validateDictionary(targetInfo, nil, requiredKeys, keyTypes, outExceptionString))
+        return false;
+
+    if (targetInfo[allFramesKey] && targetInfo[frameIDsKey]) {
+        *outExceptionString = toErrorString(nil, nil, @"The 'allFrames' and 'frameIds' properties cannot both be specified.");
+        return false;
+    }
+
+    NSArray *frameIDs = targetInfo[frameIDsKey];
+    for (NSNumber *frameID in frameIDs) {
+        if (frameID.longLongValue < mainFrameID) {
+            *outExceptionString = toErrorString(nil, frameIDsKey, @"at least one frame id is not a positive number.");
+            return false;
+        }
+    }
+
+    return true;
+}
+
+bool WebExtensionAPIScripting::validateCSS(NSDictionary *cssInfo, NSString **outExceptionString)
+{
+    static NSArray<NSString *> *requiredKeys = @[
+        targetKey,
+    ];
+    static NSDictionary<NSString *, id> *keyTypes = @{
+        cssKey: NSString.class,
+        filesKey: @[ NSString.class ],
+        targetKey: NSDictionary.class,
+    };
+
+    if (!validateDictionary(cssInfo, nil, requiredKeys, keyTypes, outExceptionString))
+        return false;
+
+    if (!validateTarget(cssInfo[targetKey], outExceptionString))
+        return false;
+
+    if ((cssInfo[cssKey] && cssInfo[filesKey])
+        || (!cssInfo[filesKey] && !cssInfo[cssKey])) {
+        *outExceptionString = toErrorString(nil, nil, @"Either 'css' or 'files' must be passed, but not both at the same time.");
```

Comment:
```
*outExceptionString = toErrorString(nil, targetKey, @"it cannot specify both 'css' and 'files'");
```

Reply:
will split this into two checks so that the error message can be `it cannot specify both 'css' and 'files'` or `either 'css' or 'files' must be passed`